tests/react/utils/useHydrateAtoms.test.tsx

```diff
@@ -1,19 +1,92 @@
 import { StrictMode, useEffect, useRef } from 'react'
 import { expect, it, jest } from '@jest/globals'
 import { fireEvent, render } from '@testing-library/react'
-import { useAtom } from 'jotai/react'
+import { useAtom, useAtomValue } from 'jotai/react'
 import { useHydrateAtoms } from 'jotai/react/utils'
 import { atom } from 'jotai/vanilla'
 
 it('useHydrateAtoms should only hydrate on first render', async () => {
   const countAtom = atom(0)
+  const statusAtom = atom('fulfilled')
+
+  const Counter = ({
+    initialCount,
+    initialStatus,
+  }: {
+    initialCount: number
+    initialStatus: string
+  }) => {
+    useHydrateAtoms([
+      [countAtom, initialCount],
+      [statusAtom, initialStatus],
+    ])
+    const [countValue, setCount] = useAtom(countAtom)
+    const [statusValue, setStatus] = useAtom(statusAtom)
 
-  const Counter = ({ initialCount }: { initialCount: number }) => {
-    useHydrateAtoms([[countAtom, initialCount]])
+    return (
+      <>
+        <div>count: {countValue}</div>
+        <button onClick={() => setCount((count) => count + 1)}>dispatch</button>
+        <div>status: {statusValue}</div>
+        <button
+          onClick={() =>
+            setStatus((status) =>
+              status === 'fulfilled' ? 'rejected' : 'fulfilled'
+            )
+          }>
+          update
+        </button>
+      </>
+    )
+  }
+  const { findByText, getByText, rerender } = render(
+    <StrictMode>
+      <Counter initialCount={42} initialStatus="rejected" />
+    </StrictMode>
+  )
+
+  await findByText('count: 42')
+  await findByText('status: rejected')
+  fireEvent.click(getByText('dispatch'))
+  fireEvent.click(getByText('update'))
+  await findByText('count: 43')
+  await findByText('status: fulfilled')
+
+  rerender(
+    <StrictMode>
+      <Counter initialCount={65} initialStatus="rejected" />
+    </StrictMode>
+  )
+  await findByText('count: 43')
+  await findByText('status: fulfilled')
+})
+
+it('useHydrateAtoms should only hydrate on first render using a Map', async () => {
+  const countAtom = atom(0)
+  const activeAtom = atom(true)
+
+  const Counter = ({
+    initialActive = false,
+    initialCount,
+  }: {
+    initialActive?: boolean
+    initialCount: number
+  }) => {
+    useHydrateAtoms(
+      new Map<
+        typeof activeAtom | typeof countAtom,
+        typeof initialActive | typeof initialCount
+      >([
```

Comment:
Note that these generics passed to `Map` were required before this PR as well, I tested it and added screenshots in the PR description showing BEFORE/AFTER.